src/intrusive_double_linked_list.rs

```diff
@@ -48,278 +50,335 @@ impl<T> DerefMut for ListNode<T> {
-        LinkedListIterator { current: self.head }
-    }
-
-    /// Consumes the list and creates an iterator over the linked list in reverse
-    /// direction.
-    /// Safety: This function is only safe as long as all pointers which are
-    /// stored inside the linked list are valid.
-    pub unsafe fn into_reverse_iter(self) -> LinkedListReverseIterator<T> {
-        LinkedListReverseIterator { current: self.tail }
-    }
-
-    /// Adds an item to the front of the linked list.
-    /// Safety: This function is only safe as long as all pointers which are
-    /// stored inside the linked list are valid.
-    pub unsafe fn add_front(&mut self, item: *mut ListNode<T>) {
-        assert!(!item.is_null(), "Can not add null pointers");
-        (*item).next = self.head;
-        (*item).prev = null_mut();
-        if !self.head.is_null() {
-            (*self.head).prev = item;
-        }
-        self.head = item;
-        if self.tail.is_null() {
-            self.tail = item;
+    /// Adds a node at the front of the linked list.
+    /// Safety: This function is only safe as long as `node` is guaranteed to
+    /// get removed from the list before it gets moved or dropped.
+    /// In addition to this `node` may not be added to another other list before
+    /// it is removed from the current one.
+    pub unsafe fn add_front(&mut self, node: &mut ListNode<T>) {
+        node.next = self.head;
+        node.prev = None;
+        match self.head {
+            Some(mut head) => head.as_mut().prev = Some(node.into()),
+            None => {}
+        };
+        self.head = Some(node.into());
+        if self.tail.is_none() {
+            self.tail = Some(node.into());
         }
     }
 
-    /// Inserts an item into the list in a way that the list keeps being sorted.
-    /// Safety: This function is only safe as long as all pointers which are
-    /// stored inside the linked list are valid.
-    pub unsafe fn add_sorted(&mut self, item: *mut ListNode<T>)
+    /// Inserts a node into the list in a way that the list keeps being sorted.
+    /// Safety: This function is only safe as long as `node` is guaranteed to
+    /// get removed from the list before it gets moved or dropped.
+    /// In addition to this `node` may not be added to another other list before
+    /// it is removed from the current one.
+    pub unsafe fn add_sorted(&mut self, node: &mut ListNode<T>)
     where
         T: PartialOrd,
     {
-        if item.is_null() {
+        if self.head.is_none() {
+            // First node in the list
+            self.head = Some(node.into());
+            self.tail = Some(node.into());
             return;
         }
 
-        if self.head.is_null() {
-            // First item in the list
-            self.head = item;
-            self.tail = item;
-            return;
-        }
-
-        let mut prev: *mut ListNode<T> = null_mut();
+        let mut prev: Option<NonNull<ListNode<T>>> = None;
```

Comment:
I'm surprised the explicit type annotation can't be inferred here!

Reply:
No, doesn't work. It doesn't understand `prev.as_mut().next = Some(node.into());` then